### DISCUSSION TITLE
window_role criteria variable.

### DISCUSSION THREAD

**Discussion by haddadnidal** [i3/i3#4660] [Q&A] [answered]
Hi, i have i3 installed and i am using Evolution as my email client. i wanted to know if the variable inside `[window_role="$variable"] ` can take just the beginning of the `wm_window_role` from the window itself and change it's behavior as i want. 
Example : My **main window** role is `EShellWindow-94918293030608` and the **Send Email window** role is `EMsgComposer-94918322275296` i want i3 to assign the `Eshell` part to Tiling and the `EMsg` window to floating.
the numbers are randomly generated by the evolution app.

Thank you.

Debian 11
i3 version 4.19.1-non-git

**Comment by Airblader** [comment, marked as answer]
Most match criteria, including `window_role`, are regular expressions, so you can do prefix / suffix matches using that. I'm not sure I really understood the question, but it seems like this might be the answer.

**Reply by haddadnidal** [reply]
so the prefix match can be done using this syntax ? `for_window [class="evolution" instance="EMsgComposer-*"] floating enable
`

**Reply by Airblader** [reply]
That's not a regular expression doing a prefix match, no. You'd need `.*` to match "anything". That shouldn't even be needed at all, though, since you don't match on line start/end, so just `EMsgComposer` should be enough.

**Reply by haddadnidal** [reply]
This is my code : `xprop | grep -i 'role'`
`WM_WINDOW_ROLE(STRING) = "EMsgComposer-94662726633504"`
I want to match the first part of the string  `EMsgComposer` since the rest is RNG.
`EMsgComposer` did not work.

Never mind it worked. Thanks alot.  I am posting my i3 code for the next guy/girl in the search of the answer. 
```
for_Window [class="Evolution" window_role="EMsgComposer"] floating enable 
for_Window [class="Evolution" window_role="EMsgComposer"] move absolute position center
```

**Reply by Airblader** [reply]
Glad it worked out. You should combine these into a single match of chained commands:

```
for_Window [class="Evolution" window_role="EMsgComposer"] floating enable, move absolute position center
```

**Reply by haddadnidal** [reply]
Thanks.

**Reply by haddadnidal** [reply]
Just to understand more the concept. the regex search if i may say so is applied on the variable in `window_role` or i can perform a regex conditions inside of the variable ? thanks for taking your time to explain.

**Reply by Airblader** [reply]
Before i3 actually parses the configuration, it does an extra pass in which it simply substitutes the variables. The resulting config is then parsed, so in terms of behavior, just imagine your variable being drop-in replaced with its value. 

Meaning: your variable can contain the regular expression syntax. In fact, even something like this is allowed:

```
-- Assuming these variables:
-- $forw = for_window
-- $float = floating enable

$forw [class="…"] $float
```

because by the time we parse the config, all variables are already substituted.

**Reply by haddadnidal** [reply]
oh ok thank you very much.
